Episode 3 (#69)
carpentries-incubator/deep-learning-intro: _episodes/03-monitor-the-model.md

---
title: "Monitor the training process"
teaching: 0
exercises: 0
questions:
- tbd
objectives:
- "Explain the importance of splitting the training data"
- "Use the data splits to plot the training process"
- "Measure the performance of your deep neural network"
- "Interpret the training plots to recognize overfitting"
- "Implement basic strategies to prevent overfitting"
- "Understand the effect of regularization techniques"

keypoints:
- "Separate training, validation, and test sets allows monitoring and evaluating your model."

> - "Dropout is a way to prevent overfitting"

Reviewer: ```suggestion
- "Batchnormalization scales the data as part of the model."
```
Dropout was removed from this episode.

|3| 	20000104 	|1 	|7 	|0.79 	|1.0262|... |
|4| 	20000105 	|1 	|5 	|0.90 	|1.0246|... |
{: .output}

### Brief exploration of the data
Let us start with a quick look at the type of features that we find in the data.
~~~
data.columns
~~~
{:.language-python}

~~~
Index(['DATE', 'MONTH', 'BASEL_cloud_cover', 'BASEL_humidity',
       'BASEL_pressure', 'BASEL_global_radiation', 'BASEL_precipitation',
       'BASEL_sunshine', 'BASEL_temp_mean', 'BASEL_temp_min',
       ...
       'STOCKHOLM_temp_min', 'STOCKHOLM_temp_max', 'TOURS_wind_speed',
       'TOURS_humidity', 'TOURS_pressure', 'TOURS_global_radiation',
       'TOURS_precipitation', 'TOURS_temp_mean', 'TOURS_temp_min',
       'TOURS_temp_max'],
      dtype='object', length=165)
~~~
{:.output}

~~~
print({x.split("_")[-1] for x in data.columns})
~~~
{:.language-python}

~~~
{'humidity', 'radiation', 'sunshine', 'gust', 'mean', 'max', 'MONTH', 'precipitation', 'pressure', 'cover', 'min', 'speed', 'DATE'}
~~~
{:.output}

## Prepare the data

### Select a subset and split into data (X) and labels (y)
The full dataset comprises 10 years (3654 days) from which we here will only select the first 3 years.

> We will then define what exactly we want to predict from this data. A very common task with weather data is to make a predicion about the weather somewhere in the future, say the next day. The present dataset is sorted by "DATE", so for each row `*i*` in the table we can pick a corresponding feature and location from row `*i+1*` that we later want to predict with our model.

Reviewer: ```suggestion
We will then define what exactly we want to predict from this data. A very common task with weather data is to make a predicion about the weather somewhere in the future, say the next day. The present dataset is sorted by "DATE", so for each row `i` in the table we can pick a corresponding feature and location from row `i+1` that we later want to predict with our model.
```

print(f"Data split into training ({X_train.shape[0]})," \
      f" validation ({X_val.shape[0]}) and test set ({X_test.shape[0]}).")
~~~
{:.language-python}

~~~
Data split into training (767), validation (164) and test set (165).
~~~
{:.output}
      
## Build a dense neural network

### Regression and classification - how to set a training goal
- Explain how to define the output part of a neural network
- What is the loss function (and which one to chose for a regression or classification task)?

In episode 2 we trained a dense neural network on a *classification task*. For this one hot encoding was used together with a Categorical Crossentropy loss function.
This measured how close the distribution of the neural network outputs corresponds to the distribution of the three values in the one hot encoding.
Now we want to work on a *regression task*, thus not prediciting the right class for a datapoint but a certain value (could in principle also be several values). In our example we want to predict the sunshine hours in Basel (or any other place in the dataset) for tomorrow based on the weather data of all 18 locations today. 

> ### Network output layer:

Reviewer: I think this header can be removed

Author: Yes, done.

> > _________________________________________________________________
> > dense_0 (Dense)              (None, 100)               16400     
> > _________________________________________________________________
> > dense_1 (Dense)              (None, 50)                5050      
> > _________________________________________________________________
> > dense_2 (Dense)              (None, 1)                 51        
> > =================================================================
> > Total params: 21,501
> > Trainable params: 21,501
> > Non-trainable params: 0
> > _________________________________________________________________
> > ~~~
> > {:.output}
> >
> > The shape of the input layer has to correspond to the number of features in our data: 163
> > 
> > The output layer here is a dense layer with only 1 node. And we here have chosen to use *no activation function*.
> > While we might use *softmax* for a classification task, here we do not want to restrict the possible outcomes for a start.
> > 
> > In addition, we have here chosen to write the network creation as a function so that we can use it later again to initiate new models.
> {:.solution}
{:.challenge}

> When compiling the model we can define a few very important aspects.

Reviewer: ```suggestion
When compiling the model we can define a few very important aspects. We will discuss them now in more detail.
```

In the given case we want to stimulate that the prodicted values are as close as possible to the true values. This is commonly done by using the *mean squared error* (mse) or the *mean absolute error* (mae), both of which should work OK in this case. Often, mse is prefered over mae because it "punishes" large prediction errors more severely.
In keras this is implemented in the `keras.losses.MeanSquaredError` class (see keras documentation: https://keras.io/api/losses/).

### Optimizer:
Somewhat coupled to the loss function is the *optimizer* that we want to use. 
The *optimizer* here refers to the algorithm with which the model learns to optimize on the set loss function. A basic example for such an optimizer would be *stochastic gradient descent*. For now, we can largely skip this step and simply pick one of the most common optimizers that works well for most tasks: the *Adam optimizer*. 

### Metrics:
In our first example (episode 2) we plotted the progression of the loss during training. 
That is indeed a good first indicator if things are working alright, i.e. if the loss is indeed decreasing as it should. 

> However, when models become more complicated then also the loss functions often become less intuitive (side remark: e.g. when adding L1 or L2 regularization). 

Reviewer: I would remove this side remark because we are not discussing L1 and L2 regularization.
```suggestion
However, when models become more complicated then also the loss functions often become less intuitive. 
```

Author: Good point, thanks.

That is why it is good practice to monitor the training process with additional, more intuitive metrics. 
They are not used to optimize the model, but are simply recorded during training. 
With Keras they can simply be added via `metrics=[...]` and can contain one or multiple metrics of interest. 
Here we could for instance chose to use `'mae'` the mean absolute error, or the the *root mean squared error* (RMSE) which unlike the *mse* has the same units as the predicted values.

~~~
model.compile(optimizer='adam',
              loss='mse',

>               metrics=[keras.metrics.RootMeanSquaredError()])

Reviewer: Does this line still have added value? As the loss is also MSE. Maybe add MAE as extra metric?

Author: The loss is `MSE`, but the metric here is `RMSE` because that has the right units... so I guess that's fine, right?

~~~
{: .language-python}

## Train a dense neural network

> Now that we created and compiled our dense neural network, we can start training it.

Reviewer: Maybe this would be a good point to explain what `batch_size` means. The concept of training batch is not explained in episode 1 and 2. It seems like an important concept to explain, also in order to understand batch normalization. This should probably also be added to the learning objectives.

Author: I added a sentence or two on it here. But I feel that might be a bit much for some... let's re-evaluate that and see where to introduce it in a proper way. I feel like gradient descent, learning rate, and batch-size all belong together and would benefit from being introduced together with a 3D loss-landscape plot.

Reviewer: I agree, I created an issue #106

> 
> * Is the performance of the model as you expected (or better/worse)? 
> * Is there a noteable difference between training set and test set? And if so, any idea why?
> > ~~~
> > y_train_predicted = model.predict(X_train)
> > y_test_predicted = model.predict(X_test)
> > ~~~
> > {: .language-python}
> > We can then compare those to the true labels, for instance by
> > ~~~
> > fig, axes = plt.subplots(1, 2, figsize=(12, 6))
> > plt.style.use('ggplot')  # optional, that's only to define a visual style
> > axes[0].scatter(y_train_predicted, y_train, s=10, alpha=0.5, color="teal")
> > axes[0].set_title("training set")
> > axes[0].set_xlabel("predicted sunshine hours")
> > axes[0].set_ylabel("true sunshine hours")
> > 
> > axes[1].scatter(y_test_predicted, y_test, s=10, alpha=0.5, color="teal")
> > axes[1].set_title("test set")
> > axes[1].set_xlabel("predicted sunshine hours")
> > axes[1].set_ylabel("true sunshine hours")
> > ~~~
> > {: .language-python}
> > ![Scatter plot to evaluate training and test set](../fig/03_regression_training_test_comparison.png)

> > > Maybe that is not exactly what you expected? What is the issue here? Any ideas?

Reviewer: I think this line can be removed
```suggestion
```

> > It is a very clear signature of **overfitting** which means that the model has to some extend memorized aspects of the training data. 
> > As a result makes much more accurate predictions on the training data than on unseen data.
> {:.solution}
{:.challenge}

Overfitting also happens in classical machine learning, but there it is usually interpreted as the model having more parameters than the training data would justify (say, a decision tree with too many branches for the number of training instances). As a consequence one would reduce the number of parameters to avoid overfitting.
In deep learning the situation is slightly different. It can -same as for classical machine learning- also be a sign of having a *too big* model, meaning a model with too many parameters (layers and/or nodes). However, in deep learning higher number of model parameters are often still considered acceptable and models often perform best (in terms of prediction accuracy) when they are at the verge of overfitting. So, in a way, training deep learning models is always a bit like playing with fire...

## Watch your model training closely

> As we just saw, deep learning models are prone to overfitting. Instead of iterating through countless cycles of model trainings and subsequent evaluations with a reserved test set, it is common practice to work with a 2nd split off dataset to monitor the model during training. This is the *validation set* which can be regarded as a 2nd test set. As with the test set the datapoints of the *validation set* are not used for the actual model training itself. Instead we evalute the model with the *validation set* after every epoch during training, for instance to splot if we see signs of clear overfitting.

Reviewer: ```suggestion
As we just saw, deep learning models are prone to overfitting. Instead of iterating through countless cycles of model trainings and subsequent evaluations with a reserved test set, it is common practice to work with a second split off dataset to monitor the model during training. This is the *validation set* which can be regarded as a second test set. As with the test set the datapoints of the *validation set* are not used for the actual model training itself. Instead we evaluate the model with the *validation set* after every epoch during training, for instance to splot if we see signs of clear overfitting.
```

- "Measure the performance of your deep neural network"
- "Interpret the training plots to recognize overfitting"
- "Implement basic strategies to prevent overfitting"
- "Understand the effect of regularization techniques"

keypoints:
- "Separate training, validation, and test sets allows monitoring and evaluating your model."
- "Dropout is a way to prevent overfitting"
---

## Import dataset
Here we want to work with the *weather prediction dataset* which can be downloaded from [ADD ZENODO LINK].
It contains daily weather observations from 18 different European cities or places through the years 2000 to 2010. For all locations the data contains the variables ‘mean temperature’, ‘max temperature’, and ‘min temperature’. In addition, for multiple of the following variables are provided: 'cloud_cover', 'wind_speed', 'wind_gust', 'humidity', 'pressure', 'global_radiation', 'precipitation', 'sunshine', but not all of them are provided for all locations. A more extensive description of the dataset including the different physical units is given in accompanying metadata file.
![18 locations in the weather prediction dataset](../fig/03_weather_prediction_dataset_map.png)

~~~

> filename_data = os.path.join(path_data, "weather_prediction_dataset.csv")

Author: Yes sounds good. I first had something like this
```python
path_root = os.path.dirname(os.getcwd())
path_data = os.path.join(path_root, "data_weather_prediction")
```
But leaving out the path is maybe better (less cognitive load).

data = pd.read_csv(filename_data)
data.head()
~~~
{:.language-python}

| | DATE 	| MONTH | 	BASEL_cloud_cover 	| 	BASEL_humidity 	| 	BASEL_pressure	| ... |
|------:|------:|---------------:|--------------:|------------------:|------------:|------------:|
|0| 	20000101 	|1 	|8 	|0.89 	|1.0286|... |
|1| 	20000102 	|1 	|8 	|0.87 	|1.0318|... |
|2| 	20000103 	|1 	|5 	|0.81 	|1.0314|... |
|3| 	20000104 	|1 	|7 	|0.79 	|1.0262|... |
|4| 	20000105 	|1 	|5 	|0.90 	|1.0246|... |
{: .output}

### Select a subset and split into data (X) and labels (y)
The full dataset comprises 10 years (3654 days) from which we here will only select the first 3 years.

> We will then define what exactly we want to predict from this data. Possible options could be to create an *interpolation* task by removing one location and predicing its values based on the remaining 17 locations. An even more frequent task when it comes to weather data, however, is to make a predicion about the weather somewhere in the future, say the next day. The present dataset is sorted by "DATE", so we can simply pick a feature and location that we want to predict with our model.

Author: I will remove the mention of the *interpolation* option since it is not relevant for the rest of the material.

> We will then define what exactly we want to predict from this data. Possible options could be to create an *interpolation* task by removing one location and predicing its values based on the remaining 17 locations. An even more frequent task when it comes to weather data, however, is to make a predicion about the weather somewhere in the future, say the next day. The present dataset is sorted by "DATE", so we can simply pick a feature and location that we want to predict with our model.

Author: And I tried to explain better how we split the data.

In episode 2 we trained a dense neural network on a *classification task*. For this one hot encoding was used together with a Categorical Crossentropy loss function.
This measured how close the distribution of the neural network outputs corresponds to the distribution of the three values in the one hot encoding.
Now we want to work on a *regression task*, thus not prediciting the right class for a datapoint but a certain value (could in principle also be several values). In our example we want to predict the sunshine hours in Basel (or any other place in the dataset) for tomorrow based on the weather data of all 18 locations today. 

### Network output layer:
The network should hence output a single float value which is why the last layer of our network will only consist of a single node. 

> ## Create the neural network
>
> We have seen how to build a dense neural network in episode 2. 
> Try now to construct a dense neural network with 3 layers for a regression task.
> You could for instance start with a network of a dense layer with 100 nodes, followed by one with 50 nodes and finally an output layer.
>
> * What must here be the dimension of our input layer?

> > * How would our output layer look like? What about the activation function?

Author: Good suggestion. Thanks!

>
> > ## Solution
> > ~~~
> > def create_nn(n_features, n_predictions):
> >     # Input layer
> >     input = Input(shape=(n_features,), name='input')
> > 
> >     # Dense layers
> >     layers_dense = Dense(100, 'relu')(input)
> >     layers_dense = Dense(50, 'relu')(layers_dense)
> > 
> >     # Output layer
> >     output = Dense(n_predictions)(layers_dense)
> > 
> >     return Model(inputs=input, outputs=output, name="weather_prediction_model")
> > 
> > model = create_nn(n_features=X_data.shape[1], n_predictions=1)

> > > model.summary()

Author: Thanks. This is also in style more consistent with episode 2 I just realized (`keras.layers.Dense`...).

### Metrics:
In our first example (episode 2) we plotted the progression of the loss during training. 
That is indeed a good first indicator if things are working alright, i.e. if the loss is indeed decreasing as it should. 
However, when models become more complicated then also the loss functions often become less intuitive (side remark: e.g. when adding L1 or L2 regularization). 
That is why it is good practice to monitor the training process with additional, more intuitive metrics. 
They are not used to optimize the model, but are simply recorded during training. 
With Keras they can simply be added via `metrics=[...]` and can contain one or multiple metrics of interest. 
Here we could for instance chose to use `'mae'` the mean absolute error, or the the *root mean squared error* (RMSE) which unlike the *mse* has the same units as the predicted values.

~~~
model.compile(optimizer='adam',
              loss='mse',
              metrics=[tf.keras.metrics.RootMeanSquaredError()])
~~~
{: .language-python}

## Train a dense neural network
Now that we created and compiled our dense neural network, we can start training it.
~~~
history = model.fit(X_train, y_train,

>                     batch_size=10,

Author: True. Good point. I was only playing around myself. We could leave it with the default value...

## Watch your model training closely
As we just saw, deep learning models are prone to overfitting. Instead of iterating through countless cycles of model trainings and subsequent evaluations with a reserved test set, it is common practice to work with a 2nd split off dataset to monitor the model during training. This is the *validation set* which can be regarded as a 2nd test set. As with the test set the datapoints of the *validation set* are not used for the actual model training itself. Instead we evalute the model with the *validation set* after every epoch during training, for instance to splot if we see signs of clear overfitting.

Let's give this a try!

We need to initiate a new model -- otherwise Keras will simply assume that we want to continue training the model we already trained above.
~~~
model = create_nn(n_features=X_data.shape[1], n_predictions=1)
model.compile(optimizer='adam',
              loss='mse',
              metrics=[tf.keras.metrics.RootMeanSquaredError()])
~~~
{: .language-python}

> But now we train it with the small addition of also passing it our validation set:

Author: Good point, was also raised by Colin. I now made issue #105 for this. Hope I soon get the time to try training on a subset, by rather would like to do this after a first merge.

~~~
history = model.fit(X_train, y_train,
                    batch_size=10,
                    epochs=200,
                    validation_data=(X_val, y_val),
                    verbose=2)
~~~
{: .language-python}

> As before the history allows plotting the training progress.

Author: Good point. Done.

~~~
history_df = pd.DataFrame.from_dict(history.history)
sns.lineplot(data=history_df[['root_mean_squared_error', 'val_root_mean_squared_error']])
plt.xlabel("epochs")
plt.ylabel("RMSE")
~~~
{: .language-python}
![Output of plotting sample](../fig/03_training_history_2_rmse.png)

This clearly shows that something is not completely right here. 
The model predictions on the validation set quickly seem to reach a plateau while the performance on the training set keeps improving.
That is a clear signature of overfitting.

## Counteract model overfitting
Overfitting is a very common issue and there are many strategies to handle it.
Most similar to classical machine learning might to **reduce the number of parameters**.

> > ## Try to reduce the degree of overfitting by lowering the number of parameters

Author: I added a suggestion for when time is short.

As the name suggests, this technique just means that you stop the model training if things do not seem to improve anymore.
More specifically, this usually means that the training is stopped if the validation loss does not (notably) improve anymore.
Early stopping is both intuitive and effective to use, so it has become a standard addition for model training.

To better study the effect, we can now savely go back to models with many (too many?) parameters:
~~~
model = create_nn(X_data.shape[1], 1, 100, 50)
model.compile(optimizer='adam',
              loss='mse',
              metrics=[tf.keras.metrics.RootMeanSquaredError()])
~~~
{: .language-python}

To apply early stopping during training it is easiest to use keras `EarlyStopping` class.
This allows to define the condition of when to stop training. In our case we will say when the validation loss is lowest.
However, since we have seen quiet some fluctuation of the losses during training above we will also set `patience=10` which means that the model will stop training of the validation loss has not gone down for 10 epochs.
~~~
from tensorflow.keras.callbacks import EarlyStopping

earlystopper = EarlyStopping(

>     monitor='val_loss', mode="min",

Author: I will remove `mode='min'` since the default is `auto` and works well in our case.

![Dropout sketch](../fig/neural_network_sketch_dropout.png)
The intuition behind dropout is that it enforces redundancies in the network by constantly removing different elements of a network. The model can no longer rely on individual nodes and instead must create multiple "paths". In addition, the model has to make predictions with much fewer nodes and weights (connections between the nodes). 
As a result, it becomes much harder for a network to memorize particular features. At first this might appear a quiet drastic approach which affects the network architecture strongly.
In practice, however, dropout is computationally a very elegant solution which does not affet training speed. And it frequently works very well.

**Important to note:** Dropout layers will only randomly silence nodes during training! During a predictions step, all nodes remain active (dropout is off).

Let's add dropout to our neural network which we will do by using keras `Dropout` layer (documentation & reference: https://keras.io/api/layers/regularization_layers/dropout/).

> One additional change that we will make here is to lower the learning rate because in the last training example the losses seemed to fluctuate a lot.

Author: Dropout was now removed entirely from this episode (now part of episode 4).

_________________________________________________________________
dropout (Dropout)            (None, 100)               0         
_________________________________________________________________
dense_13 (Dense)             (None, 50)                5050      
_________________________________________________________________
dropout_1 (Dropout)          (None, 50)                0         
_________________________________________________________________
dense_14 (Dense)             (None, 1)                 51        
=================================================================
Total params: 21,501
Trainable params: 21,501
Non-trainable params: 0
_________________________________________________________________
~~~
{: .output}

Compared to the models above, this required little changes. We add two `Dropout` layers, one after each dense layer and specify the dropout rate.
Here we use `rate=0.2` which means that at any training step 20% of all nodes will be turned off.
You can also see that Dropout layers do not add additional parameters.
Now, let's train our new model and plot the losses:

~~~
history = model.fit(X_train, y_train,

>                     batch_size = 50,

Author: No particular reason. I was playing with different batch_sizes and kept it at some value that worked (of course many other values would work as well).

>                     batch_size = 50,

Author: As above I agree that we can leave it at the default value. I will change this.

>                     batch_size = 50,

Author: I have now always set `batch_size=32`

                    verbose = 2)

history_df = pd.DataFrame.from_dict(history.history)
sns.lineplot(data=history_df[['root_mean_squared_error', 'val_root_mean_squared_error']])
plt.xlabel("epochs")
plt.ylabel("RMSE")
~~~
{: .language-python}      

![Output of plotting sample](../fig/03_training_history_5_rmse_batchnorm.png)

## Run on test set and compare to naive baseline
It seems that no matter what we add, the overall loss does not decrease much further (we at least avoided overfitting though!).
Let's again plot the results on the test set:
~~~
y_test_predicted = model.predict(X_test)

plt.figure(figsize=(5, 5), dpi=100)
plt.scatter(y_test_predicted, y_test, s=10, alpha=0.5)
plt.xlabel("predicted sunshine hours")
plt.ylabel("true sunshine hours")
~~~
{: .language-python} 

![Output of plotting sample](../fig/03_regression_test_5_dropout_batchnorm.png)

Well... certainly not perfect. But how good or bad is this? Maybe not good enough to plan your picnic for tomorrow.
But let's better compare it to a naive baseline.
Maybe the simplest prediction to make would be to say: Tomorrow we will have the same number of sunshine hours as today. Let's compare to this.

> ~~~

Reviewer: I would make an exercise of this. Ask them to plot the baseline, and calculate the RMSE (probably need to refer to https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_squared_error.html ) to compare

plt.figure(figsize=(5, 5), dpi=100)
plt.scatter(X_test["BASEL_sunshine"], y_test, s=10, alpha=0.5)
plt.xlabel("sunshine hours yesterday")
plt.ylabel("true sunshine hours")
~~~
{: .language-python} 

![Output of plotting sample](../fig/03_regression_test_5_naive_baseline.png)

> ## Optional part - prediction uncertainty using Monte-Carlo Dropout

Reviewer: I haven't reviewed this part in detail yet, I think it is really cool but maybe should go in a separate (optional) episode

Author: I think I agree. I mostly wanted to have this for the course this week...

Author: I now moved the optional part to a separate .md file.

Our models get the general trends right, but still predictions vary quiet a bit and can even be far off.

> ## Open question: What could be next steps to further improve the model?
> 
> With unlimited options to modify the model architecture or to play with the training parameters, deep learning can trigger very extensive hunting for better and better results.
> Usually models are "well behaving" in the sense that small chances to the architectures also only result in small changes of the performance (if any).
> It is often tempting to hunt for some magical settings that will lead to much better results. But do those settings exist?
> Applying common sense is often a good first step to make a guess of how much better *could* results be. 
> In the present case we might certainly not expect to be able to reliably predict sunshine hours for the next day with 5-10 minute precision.
> But how much better our model could be exactly, often remains difficult to answer.
> 
> * What changes to the model architecture might make sense to explore?
> * Ignoring changes to the model architecture, what might notably improve the prediction quality?

> > 
> > > This is on open question. And we don't actually know how far one could push this sunshine hour preciction (try it out yourself if you like! We're curious!).

Reviewer: ```suggestion
> 
> > # Solution
> > This is on open question. And we don't actually know how far one could push this sunshine hour prediction (try it out yourself if you like! We're curious!).
```

~~~
{:.output}
      
## Build a dense neural network

### Regression and classification - how to set a training goal

> - Explain how to define the output part of a neural network

Author: Good point. I added this to the objectives.

> What might be a bit unintuitive is that the training runs might now end very rapidly (in particular when the learning rate is high).

Reviewer: ```suggestion
What might be a bit unintuitive is that the training runs might now end very rapidly
```

We haven't introduced the learning rate (yet) so cannot refer to it.

> ~~~

Reviewer: This needs to be formatted as an exercise, and a bit of introduction, such as:
```suggestion
Look at the [documentation of the batch normalization layer](https://keras.io/api/layers/normalization_layers/batch_normalization/). Add this as a first layer to the model we defined above. Train the model and compare the performance to the model without batch normalization.
~~~
```

Author: Sounds good to me. Thanks for adding this!

> ## Exercise: Add a BatchNormalization layer as the first layer to your neural network.

Reviewer: This is not yet formatted as exercise

Author: Formatted this as exercise now.

    # Dense layers
    layers_dense = keras.layers.BatchNormalization()(layers_input)
    layers_dense = keras.layers.Dense(100, 'relu')(layers_dense)
    layers_dense = keras.layers.Dense(50, 'relu')(layers_dense)

    # Output layer
    layers_output = keras.layers.Dense(n_predictions)(layers_dense)

    # Defining the model and compiling it
    return keras.Model(inputs=layers_input, outputs=layers_output, name="model_batchnorm")

model = create_nn(X_data.shape[1], 1)

> model.compile(loss='mse', optimizer=Adam(1e-4), metrics=[keras.metrics.RootMeanSquaredError()])

Reviewer: Do we need to specify learning rate here? In the previous code snippets we did `optimizer='adam'`, and the learning rate is not yet explained.

Author: True, that's historical I guess.... I set it to the simple default `adam`

Well... certainly not perfect. But how good or bad is this? Maybe not good enough to plan your picnic for tomorrow.
But let's better compare it to a naive baseline.

> > # Create a similar scatter plot as above for a reasonable baseline
> >
> > What can we take as a baseline? 
> > Maybe the simplest prediction to make would be to say: Tomorrow we will have the same number of sunshine hours as today.
> > Let's compare to this.
> > 
> > > # Solution
> > > We can here just take the `BASEL_sunhine` column of our data, because this contains the sunshine hours from one day before what we have as a label.
> > > ~~~
> > > plt.figure(figsize=(5, 5), dpi=100)
> > > plt.scatter(X_test["BASEL_sunshine"], y_test, s=10, alpha=0.5)
> > > plt.xlabel("sunshine hours yesterday")
> > > plt.ylabel("true sunshine hours")
> > > ~~~
> > > {: .language-python} 
> > > 
> > > ![Output of plotting sample](../fig/03_regression_test_5_naive_baseline.png)

Reviewer: ```suggestion
> ## Exercise: Create a similar scatter plot as above for a reasonable baseline
>
> What can we take as a baseline? 
> Maybe the simplest prediction to make would be to say: Tomorrow we will have the same number of sunshine hours as today.
> Let's compare to this.
> 
> > ## Solution
> > We can here just take the `BASEL_sunhine` column of our data, because this contains the sunshine hours from one day before what we have as a label.
> > ~~~
> > plt.figure(figsize=(5, 5), dpi=100)
> > plt.scatter(X_test["BASEL_sunshine"], y_test, s=10, alpha=0.5)
> > plt.xlabel("sunshine hours yesterday")
> > plt.ylabel("true sunshine hours")
> > ~~~
> > {: .language-python} 
> > 
> > ![Output of plotting sample](../fig/03_regression_test_5_naive_baseline.png)
> {:.solution}
{:.challenge}
```

> * What changes to the model architecture might make sense to explore?
> * Ignoring changes to the model architecture, what might notably improve the prediction quality?
> 

> > > # Solution
> > > This is on open question. And we don't actually know how far one could push this sunshine hour prediction (try it out yourself if you like! We're curious!).

Reviewer: ```suggestion
> > ## Solution
> > This is on open question. And we don't actually know how far one could push this sunshine hour prediction (try it out yourself if you like! We're curious!).
```
